include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -118,9 +126,53 @@ namespace SampleFlow
                 const unsigned int max_delays,
                 const types::sample_index n_samples,
                 const std::mt19937::result_type random_seed = {});
+      private:
+        /**
+         * Recursively compute the acceptance ratio given the previous sample and
+         * a vector of proposed samples (all but one that have already been rejected).
+         *
+         * @param[in] x The previous accepted sample.
+         * @param[in] y Vector of proposed samples; all but the last element have already
+         *   been rejected.
+         * @param[in] log_likelihood The ratio of the likelihoods of x and the current
+         *   proposed sample (the last element in y).
+         */
+        double
+        alpha_fn (const OutputType &x,
+                  const std::vector<OutputType> &y,
+                  const std::function<double (const OutputType &)> &log_likelihood);
     };
 
 
+    template <typename OutputType>
+    double 
+    DelayedRejectionMetropolisHastings<OutputType>::
+    alpha_fn (const OutputType &x,
+              const std::vector<OutputType> &y,
+              const std::function<double (const OutputType &)> &log_likelihood)
+    {
+        // define the current delay stage as the number of rejected samples
+        // (number of proposed samples minus 1)
+        const int i = y.size() - 1;
+        // get the current proposed sample from the list of all proposed samples
+        const OutputType yi = y[i];
+        // get the likelihood ratio of the previous sample to the current proposed sample
+        const double likelihood_ratio = std::exp(log_likelihood(x) - log_likehood(yi));
+        // in the case where no samples have been delayed, the acceptance ratio is calculated
+        // the same as regular MH; we assume that the proposal is symmetric, so the acceptance
+        // ratio is simply the likelihood ratio
+        if (i == 0)
+            return likelihood_ratio;
+        // otherwise, recursively compute acceptance ratio
+        double alpha = likelihood_ratio;
+        for (j = 0; j < y.size()) {
+            alpha *= (1 - alpha_fn(yi, y[i - j:i:-1], log_likelihood)) /
+                     (1 - alpha_fn(x, y[:j], log_likelihood));
```

Comment:
Here, I have the desired "slicing" behavior written in the python syntax that I'm familiar with. @bangerth , do you know an easy way to replicate this with C++ vectors? The first example, `y[i - j:i:-1]` means, take elements from `y`, starting at index `i - j` (inclusive), ending at index `i` (exclusive), and in reverse. The second example `y[:j]` just means take the first `j` elements.

Reply:
You could do something like this:
```suggestion
            alpha *= (1 - alpha_fn(yi, std::vector<double>(y.begin()+i - j, y.begin()+i]), log_likelihood)) /
                     (1 - alpha_fn(x, std::vector(y.begin(), y.begin()+j), log_likelihood));
```
but this doesn't reverse the first place. For that, you could think about using `std::vector::rbegin()/rend()`. You'd have to work out the correct `+i` or `+(i-1)` or `+(i-i)`, though.